### PR TITLE
[4.1][Feature][WIP] enableServerSideExport

### PR DESCRIPTION
**Pull request by baxri**
Simple add enableServerSideExport button Filters bar.

### CODE REVIEW

File: src/CrudPanel.php [hunk 1]
```diff
@@ -266,4 +266,13 @@ private function getRelationModel($relationString)
 
         return get_class($result);
     }
+
+    public function enableServerSideExport()
```
Comment (lloy0076) [inline src/CrudPanel.php]: Need PHP doc.

File: src/CrudServiceProvider.php [hunk 2]
```diff
@@ -76,17 +77,22 @@ public function register()
             return new CRUD($app);
         });
 
+//        dd(ExcelServiceProvider::class);
```
Comment (lloy0076) [inline src/CrudServiceProvider.php]: Please remove.

File: src/CrudServiceProvider.php [hunk 3]
```diff
@@ -76,17 +77,22 @@ public function register()
...
+
         // register its dependencies
         $this->app->register(\Backpack\Base\BaseServiceProvider::class);
         $this->app->register(\Collective\Html\HtmlServiceProvider::class);
         $this->app->register(\Barryvdh\Elfinder\ElfinderServiceProvider::class);
         $this->app->register(\Intervention\Image\ImageServiceProvider::class);
 
+        $this->app->register(ExcelServiceProvider::class);
```
Comment (lloy0076) [inline src/CrudServiceProvider.php]: Given the other registeres are using the fully qualified version (which I don't like but that's how things are), this one probably should too for consistency.

File: src/CrudServiceProvider.php [hunk 4]
```diff
@@ -76,17 +77,22 @@ public function register()
...
+
         // register their aliases
         $loader = \Illuminate\Foundation\AliasLoader::getInstance();
         $loader->alias('Form', \Collective\Html\FormFacade::class);
         $loader->alias('Html', \Collective\Html\HtmlFacade::class);
         $loader->alias('Image', \Intervention\Image\Facades\Image::class);
+//        $loader->alias('Excel', Excel::class);
```
Comment (lloy0076) [inline src/CrudServiceProvider.php]: Not needed?

File: src/app/Http/Controllers/CrudFeatures/AjaxTable.php [hunk 5]
```diff
@@ -11,6 +14,30 @@
      */
     public function search()
     {
+        $requestType = Request()->request_type;
+
+        if ($requestType == 'excel') {
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudFeatures/AjaxTable.php]: `===` is more stringent.

File: src/app/Http/Controllers/CrudFeatures/AjaxTable.php [hunk 6]
```diff
@@ -11,6 +14,30 @@
...
+            $filename = ucfirst($this->crud->entity_name_plural . '-' . Carbon::now()->toDateString());
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudFeatures/AjaxTable.php]: Consider:

```
$filename = $this->getExcelFileName(); // Or some sensible name

...

protected function getExcelFileName()
{
  return ucfirst...
}
```

...allowing the end user to more easily select the filename.

File: src/app/Http/Controllers/CrudFeatures/AjaxTable.php [hunk 7]
```diff
@@ -11,6 +14,30 @@
...
+            $result = $this->crud->query->get();
+
+            $data = array();
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudFeatures/AjaxTable.php]: `$data = [];`

File: src/app/Http/Controllers/CrudFeatures/AjaxTable.php [hunk 8]
```diff
@@ -23,17 +50,17 @@ public function search()
             $filteredRows = $this->crud->count();
         }
 
-        // start the results according to the datatables pagination
+// start the results according to the datatables pagination
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudFeatures/AjaxTable.php]: No need to change this.

File: src/app/Http/Controllers/CrudFeatures/AjaxTable.php [hunk 9]
```diff
@@ -23,17 +50,17 @@ public function search()
...
         if ($this->request->input('start')) {
             $this->crud->skip($this->request->input('start'));
         }
 
-        // limit the number of results according to the datatables pagination
+// limit the number of results according to the datatables pagination
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudFeatures/AjaxTable.php]: ditto

File: src/app/Http/Controllers/CrudFeatures/AjaxTable.php [hunk 10]
```diff
@@ -23,17 +50,17 @@ public function search()
...
         if ($this->request->input('length')) {
             $this->crud->take($this->request->input('length'));
         }
 
-        // overwrite any order set in the setup() method with the datatables order
+// overwrite any order set in the setup() method with the datatables order
```
Comment (lloy0076) [inline src/app/Http/Controllers/CrudFeatures/AjaxTable.php]: ditto

File: src/resources/views/filters/export.blade.php [hunk 11]
```diff
@@ -0,0 +1,89 @@
+<li
+        class="{{ $filter->name }}-class btn btn-default"
+        style="background-color: seagreen; color: white;   height: 32px;line-height: 17px;margin: 0px 10px; "
+        name="{{ $filter->name }}"
+        id="export-button"
+>Exel Export</li>
```
Comment (lloy0076) [inline src/resources/views/filters/export.blade.php]: Spelling! 👎 

Consider putting this into i18n too.

File: src/resources/views/filters/export.blade.php [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+<li>
+    <a href="#"
+       class="btn btn-default"
+       style="display: none; background-color: lightgrey; color: black;  margin: 0px 10px; height: 32px; line-height: 17px;"
+       id="export-button-download"
+    >Download</a>
```
Comment (lloy0076) [inline src/resources/views/filters/export.blade.php]: i18n

File: src/resources/views/filters/export.blade.php [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+        function make_server_side_ajax_export(url) {
+
+            $("#export-button-download").hide();
+
+            var button = $("#export-button");
+            var text = button.text();
+
+            button.html("");
+            button.html("Loading...");
```
Comment (lloy0076) [inline src/resources/views/filters/export.blade.php]: i18n

File: src/resources/views/filters/export.blade.php [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+
+                    if (data.error.length > 0) {
+                        alert(data.error);
+                    } else {
+
+                        $("#export-button-download").attr("href", data.download);
+                        $("#export-button-download").fadeIn();
+
+                        //window.location.href = data.download;
```
Comment (lloy0076) [inline src/resources/views/filters/export.blade.php]: Why's that here?

File: src/resources/views/filters/export.blade.php (hunk 12 above)
Comment (baxri) [reply src/resources/views/filters/export.blade.php]: do I need to translate in all languages?

Comment (lloy0076) [reply src/resources/views/filters/export.blade.php]: Only `en` - it's quite probably there's already a "download" and "loading" one about the place.

**Comment by baxri** [comment]
@lloy0076  think I done everything but I dont undertsand 

what the problem in 

continuous-integration/travis-ci/pr — The Travis CI build could not complete due to an error
continuous-integration/styleci/pr — The StyleCI analysis has failed - 3 files need addressing

**Comment by lloy0076** [comment]
Travis is failing because something requires a higher version of PHP; I'm not sure what the styleCI is complaining about (but If ind its current setting somewhat obtues).

File: src/CrudPanel.php [hunk 15]
```diff
@@ -267,12 +269,27 @@ private function getRelationModel($relationString)
         return get_class($result);
     }
 
-    public function enableServerSideExport()
+    /**
+     * Enables server side export button in filters bar
+     *
+     * @param null $filename
```
Comment (lloy0076) [inline src/CrudPanel.php]: `string|null $filename`

File: src/CrudPanel.php [hunk 16]
```diff
@@ -267,12 +269,27 @@ private function getRelationModel($relationString)
...
+     */
+    public function enableServerSideExport($filename = null)
```
Comment (lloy0076) [inline src/CrudPanel.php]: `string $filename = null)`

File: src/CrudPanel.php [hunk 17]
```diff
@@ -274,7 +274,7 @@ private function getRelationModel($relationString)
      *
      * @param string|null $filename  
      */
-    public function enableServerSideExport($filename = null)
+    public function enableServerSideExport( string $filename = null)
```
Comment (lotarbo) [inline src/CrudPanel.php]: mb (?string $filename = null ) ?

**Comment by baxri** [comment]
Travis check done 😄 
Now I think every thing is ok @lloy0076 
?

**Comment by baxri** [comment]
@lloy0076 @lotarbo is there any problems in pull request? :)

**Comment by baxri** [comment]
## All checks have passed 
3 successful checks

㊗️ 🎂 🎈 🥇 😀 !!!

**Comment by baxri** [comment]
Any news to approve this changes?

**Comment by lloy0076** [comment]
Let's see what @tabacitu thinks of this.

**Comment by b8ne** [comment]
I think a few things need to be looked at here.
1. The `exports` folder should be customisable in case my folder structure is different and I don't want a new folder sitting in my public path, perhaps add a setting to the config file?
2. If adding an 'export all columns not just those visible' feature, shouldn't it work for all export types? i.e. pdf, csv etc. not just Excel?
3. Forcing the export to render ALL columns may be just as restrictive to the developer/user as forcing to render visible Datatable columns.  We should be able to at least have the option to choose which columns are included.

I have already submitted a [PR](https://github.com/Laravel-Backpack/CRUD/pull/960) a while ago which covers the above points.   I like some of the code in this PR though, perhaps a hybrid of the 2 would be best to cover all bases?

**Comment by baxri** [comment]
I set upload path in config. "backpack.crud" is it ok?

**Comment by chemitaxis** [comment]
Hi, any news about it? Thanks...

**Comment by chemitaxis** [comment]
Hi @tabacitu I would like to know the state of this task... can I help with something? Thanks.

**Comment by tabacitu** [comment]
Hi @chemitaxis - I haven't had a chance to look at this yet, unfortunately. I haven't given it priority in the past because there was always something more urgent, or something more important, but... I do think it's time to move away from DataTables' export, since it's so freaking slow for big data sets, and it only exports the currently visible rows. I'll take a look shortly and tell you what I think. Please do the same.

@baxri - so _so_ sorry for not taking a look at this earlier. I really do appreciate the work you've put in, and realise it's frustrating to not have this merged for so long. It's not an excuse, but keep in mind I myself have PRs that have been closed, or are pending for months (even years). Not easy managing an open project, with so many moving pieces and opinions. Sometimes opinions and best practices change.

@baxri - could you please tell me what your intention with this PR was? What did you aim to solve in the existing export implementation? Only exporting the visible rows? Slow export for big data sets? Excel not formatted right? Thanks!

**Comment by chemitaxis** [comment]
Hi @tabacitu totally agree, the export client is very slow... do you have planned enable a server side rendering excel? Thanks

**Comment by ziming** [comment]
Personally I think this PR is a lot more important than it is. As it's pretty easy to have more than 1000 records in the span of a year and you will need the server side export feature to export without using too much memory.

Currently there is both Laravel Excel and spatie/laravel-simple-excel that could do this job.

https://github.com/maatwebsite/Laravel-Excel

https://github.com/spatie/simple-excel

So we might want to re-evaluate which package to use for backpack v4 server side export feature.

**Comment by smurfiez** [comment]
Since this has been sitting here for quite awhile, how can I implement this without breaking old stuff?

**Comment by tabacitu** [comment]
@ziming I agree. I think the server-side Export functionality is long overdue. And it will be particularly helpful in order to migrate away from DataTables.

@smurfiez you're right. This PR is for Backpack 3.x, but you're probably using Backpack 4.0 - so it won't work as-is. Ideally we'll get to refactor this PR into a separate Operation (Export), but I can't promise anything regarding the timeline - we're currently handling some features that take priority. If you want to do it yourself, this is how you can [create a new operation](https://backpackforlaravel.com/docs/4.0/crud-operations#creating-a-custom-operation). The way I see it, it would be something like:

```php
<?php

namespace App\Http\Controllers\Admin\Operations;

use Illuminate\Support\Facades\Route;

trait ExportOperation
{
    /**
     * Define which routes are needed for this operation.
     *
     * @param string $segment    Name of the current entity (singular). Used as first URL segment.
     * @param string $routeName  Prefix of the route name.
     * @param string $controller Name of the current CrudController.
     */
    protected function setupCommentRoutes($segment, $routeName, $controller)
    {
        Route::get($segment.'/export/pdf', [
            'as'        => $routeName.'.exportPdf',
            'uses'      => $controller.'@exportPdf',
            'operation' => 'export',
        ]);
        Route::get($segment.'/export/xls', [
            'as'        => $routeName.'.exportXls',
            'uses'      => $controller.'@exportXls',
            'operation' => 'export',
        ]);
        Route::get($segment.'/export/csv', [
            'as'        => $routeName.'.exportCsv',
            'uses'      => $controller.'@exportCsv',
            'operation' => 'export',
        ]);
    }

    /**
     * Add the default settings, buttons, etc that this operation needs.
     */
    protected function setupExportDefaults()
    {
        $this->crud->allowAccess('export');

        $this->crud->operation('export', function () {
            $this->crud->loadDefaultOperationSettingsFromConfig();
        });

        $this->crud->operation('list', function () {
            $this->crud->addButton('bottom', 'export', 'view', 'crud::buttons.export');
        });
    }

    public function exportPdf()
    {
        $this->crud->hasAccessOrFail('comment');

        // prepare the fields you need to show
        $this->data['crud'] = $this->crud;
        
        // TODO: put together the file using a library (consider using generators to be able to handle big files)
        // TODO: download the file
    }

    public function exportCsv()
    {
        $this->crud->hasAccessOrFail('comment');

        // prepare the fields you need to show
        $this->data['crud'] = $this->crud;
        
        // TODO: put together the file using a library (consider using generators to be able to handle big files)
        // TODO: download the file
    }

    public function exportXls()
    {
        $this->crud->hasAccessOrFail('comment');

        // prepare the fields you need to show
        $this->data['crud'] = $this->crud;
        
        // TODO: put together the file using a library (consider using generators to be able to handle big files)
        // TODO: download the file
    }
}
```

**Comment by tabacitu** [comment]
I have great news! We're finally working on a solution for this, Pedro is now polishing an ExportOperation that will solve this problem. Exports with millions of records... yes please! Stay tuned!

@baxri sorry your PR did not get merged. I have hundreds that didn't too... please know that your time was NOT wasted. This enabled us to think about the problem, other devs to express their desire for this, etc. Thank you for taking the time to help with this, everyone 🙏

**Pull request closed without merging**